pkg/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,76 @@
+package dispatcher
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	admissionctl "sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	responsehelper "github.com/openshift/managed-cluster-validating-webhooks/pkg/helpers"
+	"github.com/openshift/managed-cluster-validating-webhooks/pkg/webhooks"
+	"github.com/openshift/managed-cluster-validating-webhooks/pkg/webhooks/utils"
+)
+
+var log = logf.Log.WithName("dispatcher")
+
+// Dispatcher struct
+type Dispatcher struct {
+	hooks *map[string]webhooks.WebhookFactory // uri -> hookfactory
+	mu    sync.Mutex
+}
+
+// NewDispatcher new dispatcher
+func NewDispatcher(hooks webhooks.RegisteredWebhooks) *Dispatcher {
+	hookMap := make(map[string]webhooks.WebhookFactory)
+	for _, hook := range hooks {
+		hookMap[hook().GetURI()] = hook
+	}
+	return &Dispatcher{
+		hooks: &hookMap,
+	}
+}
+
+// HandleRequest http request
+func (d *Dispatcher) HandleRequest(w http.ResponseWriter, r *http.Request) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	log.Info("Handling request", "request", r.RequestURI)
+
+	// is it one of ours?
+	if hook, ok := (*d.hooks)[r.RequestURI]; ok {
+		// it's one of ours, so let's attempt to parse the request
+		request, _, err := utils.ParseHTTPRequest(r)
+		if err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			log.Error(err, "Error parsing HTTP Request Body")
+			responsehelper.SendResponse(w, admissionctl.Errored(http.StatusBadRequest, err))
+			return
+		}
+		if !hook().Validate(request) {
+			w.WriteHeader(http.StatusBadRequest)
+			responsehelper.SendResponse(w,
+				admissionctl.Errored(http.StatusBadRequest,
+					fmt.Errorf("Not a valid webhook request")))
+			return
+		}
+
+		// Dispatch
+		resp := hook().Authorized(request)
+
+		// This is optional, so if it wasn't set, we'll not do anything
+		if resp.AdmissionResponse.Result.Code != 0 {
```

Comment:
Interesting, had to do some digging into https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L656 to understand that comment.
Did you identify any occasion where this Code comes empty?

Reply:
No, but it's an optional field in the API and so I thought it best to account for that here.